### PR TITLE
[S] Allow validation of players for everybody

### PR DESCRIPTION
**Pull request by Benau**
Now when client make a join key, than server will poll and return everything needed forum same table and clear the returned data.

The v3_server_conn will auto clear some data anyway when the server is taken down (including entering the game), so more more stopper is needed.

### CODE REVIEW

**Comment by leyyin** [comment]
What is the corresponding branch in stk-code for the changes here? It is `aes` ?

**Comment by Benau** [comment]
Yes

File: include/ClientSession.class.php [hunk 1]
```diff
@@ -305,10 +305,10 @@ public function getServerConnectionRequests($ip, $port, $current_players)
                 ]);
 
             $connection_requests = DBConnection::get()->query(
-                "SELECT s.user_id, c.ip, c.private_port, c.port
-                FROM `{DB_VERSION}_server_conn` s
-                INNER JOIN `{DB_VERSION}_client_sessions` c ON c.uid = s.user_id
-                WHERE s.`server_id` = :server_id AND s.`is_request` = '1'",
+                "SELECT `user_id`, `server_id`, `ip`, `port`, `aes_key`, `iv`, `username`
+                FROM `{DB_VERSION}_server_conn` INNER JOIN `{DB_VERSION}_users`
```
Comment (leyyin) [inline include/ClientSession.class.php]: New line before INNER JOIN

File: install/install.sql [hunk 2]
```diff
@@ -344,9 +341,12 @@ CREATE TABLE IF NOT EXISTS `v3_servers` (
...
 CREATE TABLE IF NOT EXISTS `v3_server_conn` (
-    `user_id`    INT UNSIGNED NOT NULL,
-    `server_id`  INT UNSIGNED NOT NULL,
-    `is_request` BOOL         NOT NULL DEFAULT '1',
+    `user_id`    INT UNSIGNED      NOT NULL,
+    `server_id`  INT UNSIGNED      NOT NULL,
+    `ip`         INT UNSIGNED      NOT NULL,
+    `port`       SMALLINT UNSIGNED NOT NULL,
+    `aes_key`    CHAR(24)          NOT NULL,
+    `iv`         CHAR(24)          NOT NULL,
```
Comment (leyyin) [inline install/install.sql]: Maybe this should be name `aes_iv`

File: include/ClientSession.class.php [hunk 3]
```diff
@@ -200,45 +195,50 @@ public function getNotifications()
...
+     * @param int $address ip of client
+     * @param int $port port of client
+     * @param char $key aes 128 bit key in base64 of client
+     * @param char $iv initialization vector of the aes key
      *
      * @return array the ip and port of the player
      * @throws ClientSessionException if the request fails
      */
-    public function getPeerAddress($peer_id)
+    public function setJoinServerKey($server_id, $address, $port, $key, $iv)
```
Comment (leyyin) [inline include/ClientSession.class.php]: `$key` should be `$aes_key` and and `$iv` should be `$aes_iv`. Also update comments

File: api/server.php [hunk 4]
```diff
@@ -171,47 +106,28 @@
...
             try
             {
                 $userid = isset($_POST['userid']) ? (int)utf8_encode($_POST['userid']) : null;
                 $token = isset($_POST['token']) ? utf8_encode($_POST['token']) : null;
-                $server_id = isset($_POST['server_id']) ? (int)utf8_encode($_POST['server_id']) : null;
+                $server_id = isset($_POST['server-id']) ? (int)utf8_encode($_POST['server-id']) : null;
+                $address = isset($_POST['address']) ? utf8_encode($_POST['address']) : null;
+                $port = isset($_POST['port']) ? utf8_encode($_POST['port']) : null;
+                $aes_key = isset($_POST['aes-key']) ? utf8_encode($_POST['aes-key']) : null;
+                $iv = isset($_POST['iv']) ? utf8_encode($_POST['iv']) : null;
```
Comment (leyyin) [inline api/server.php]: `iv` should be `aes-iv`

**Comment by Benau** [comment]
If you have time, can u make "Kobato" and "deve" account server_hoster so they can host official server?

**Comment by leyyin** [comment]
Looks good, I will push to production once the stk-code branch is in master.
Sure I can add those roles.

**Comment by Benau** [comment]
Thank you

**Pull request merged**